### PR TITLE
emacs: add withSystemd option

### PR DESCRIPTION
**Pull request by ghost**
This patch causes emacs to honor the `systemdSupport` and `dbusSupport` attrs.

The default behavior (i.e. if these attrs are undefined) is unchanged.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [x] powerpc64le-linux
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by adisbladis** [comment]
What's the motivation for this change?

**Comment by a-m-joseph** [comment]
The motivation for this change is to have emacs/generic.nix honor the systemdSupport and dbusSupport attrs, like many other derivations do.

Emacs has compile-time logic to support building with or without each of these packages.  This patch exposes that functionality to nix users, doing so via widely-standardized attrs which can be set at the nixpkgs-wide level -- for example, in ~/.config/nixpkgs/config.nix.

**Comment by a-m-joseph** [comment]
Latest push adds a commit message, no other changes.

**Comment by adisbladis** [comment]
> The motivation for this change is to have emacs/generic.nix honor the systemdSupport and dbusSupport attrs, like many other nixpkgs expressions do.

I understand what the effect is, but what is the motivation? What makes you want to add these?

**Comment by a-m-joseph** [comment]
> > The motivation for this change is to have emacs/generic.nix honor the systemdSupport and dbusSupport attrs, like many other nixpkgs expressions do.
> 
> I understand what the effect is, but what is the motivation? What makes you want to add these?

Because I compile emacs without linking to systemd or dbus, and I use nixpkgs to do so.

I'm sure the facts in the previous sentence were already obvious to you, so clearly I have misunderstood your question, and its relevance to the merge decision.

Perhaps you will find [this comment](https://github.com/NixOS/nixpkgs/pull/160972#issuecomment-1047268404) and the commit message for #160972 to be interesting?

**Comment by a-m-joseph** [comment]
@adisbladis, I have reduced the scope of this PR to deal only with systemd and removed the dbus-related parts.  Does this enhance its mergeability?

If you need further context, please see also [this comment](https://github.com/NixOS/nixpkgs/pull/160972#issuecomment-1047268404), the [commit message for #160972](https://github.com/NixOS/nixpkgs/pull/160972/commits/07bedf0159b59b636b8e42e1d0ae05f3d9b31b59), and [this comment](https://github.com/NixOS/nixpkgs/pull/164156#issuecomment-1070384547).

**Comment by a-m-joseph** [comment]
Ping...

**Comment by peterhoeg** [comment]
I find this rather pointless TBH. 

We currently have 546 files in `pkgs` that reference `systemd`. We then have 4 files that reference `useSystemd`, 14 that reference `withSystemd`, 15 with `systemdSupport` and 9 with `enableSystemd` (separate note - we really should standardize on parameter naming) and of course it's a deeply embedded part of NixOS. 

So the only thing we gain by making this configurable is catering to the presumably tiny subset of users who are on a non-systemd distribution who is only using emacs and a no other package that pulls in systemd and who simply cannot stand the systemd binaries existing on their systems.

For the record, I don't think we should have taken the referenced patches either. It's not about being for or against systemd at all. There are some libraries that we will expect being present and that various pieces of software will build against. This is just one of them.

I'd argue that the vast majority of users simply don't care either way as long as emacs works. We shouldn't be adding flags unless there's a reasonable expectation that the non-default value will be actively used.

File: pkgs/applications/editors/emacs/generic.nix [hunk 1]
```diff
@@ -36,6 +35,8 @@
   else if withMotif then "motif"
   else if withAthena then "athena"
   else "lucid")
+, systemdSupport ? stdenv.isLinux
+, systemd ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/emacs/generic.nix:39]:
```suggestion
, systemd
```

**Review by SuperSandro2000 — COMMENTED** [review]
Is systemd already building on mips64/powerpc? If not that could maybe be a reason to have such a flag.

**Comment by a-m-joseph** [comment]
Alpine Linux is a very popular non-systemd distribution, which [supports powerpc64le](https://wiki.alpinelinux.org/wiki/Architecture).

I am going to try migrating to it as a "userspace bootloader" for nixpkgs instead of the hand-rolled equivalent that I am currently using.

**Comment by SuperSandro2000** [comment]
> The idea is that if `/nix/store` gets messed up you can always fall back to Alpine in order to fix things, sort of like the initrd shell (which is also busybox).

The nixos way to do this is to boot into a nixos live iso and reinstall the entire system or use nixos-enter if that just works.

**Comment by amjoseph-nixpkgs** [comment]
> The nixos way to do this is to boot into a nixos live iso

Except that you can't fit a spare nixos live iso into a 16Mbyte SPI flash chip. :smiley:

**Comment by amjoseph-nixpkgs** [comment]
Rebased again.

**Comment by amjoseph-nixpkgs** [comment]
Rebased yet again.

**Review by Atemu — APPROVED** [review]
As long as we're not carrying heavy patches and what not, I don't see why we shouldn't offer the option to build without systemd.

I also don't buy the argument that, because other packages don't have the option (or it's named differently), we shouldn't allow single packages to make it optional.  
How else would we get to the point where all of those packages have the option? In one big chunk? That's one heck of an effort, would take forever to merge and would probably just bitrot.  
Names are trivial to change later when we get to a point where we standardise.

I'm also squarely in the camp that most people who don't want systemd on their system are misguided hipsters but I also acknowledge that there are valid use-cases where systemd simply isn't an option or a waste of space.    
You wouldn't want to include systemd when building an OCI container for example. That'd just be wasteful. This is totally a supported use-case in Nixpkgs (there's even a ASCIInema on the website IIRC) amd we should cater towards it if people are willing to maintain it or it's trivial to maintain for the rest of us.

@amjoseph-nixpkgs If you can gather a team of people who need/want packages without systemd who would be willing to maintain them, I wouldn't mind heavier patches than this either.

Btw, as a hack, you could also try using an overlay where `systemd = null;`. That should get filtered out of any buildInputs automatically.

**Comment by amjoseph-nixpkgs** [comment]
> @amjoseph-nixpkgs If you can gather a team of people who need/want packages without systemd who would be willing to maintain them

All of them run Guix.

I'm the freak case who can't stand scheme and is thinks nixlang so beautiful that I'm willing to build everything from source and carry (previously) a dozen out-of-tree patches.  This is the *last one* that isn't merged.  I'm sure there are plenty of people interested in using nixpkgs on Linux without systemd but uneasy about having to carry out-of-tree patches.

I understand why NixOS has a hard dependency on systemd and am opposed to changing that.  The Nix community doesn't have the skillset to do what Guix, Void, Alpine, and Gentoo are doing.  But I also am opposed to nixpkgs depending on systemd.

> I wouldn't mind heavier patches than this either.

I would mind that.  None of my `systemdSupport` requires patches of any kind, and I intend to keep it that way.

**Comment by Atemu** [comment]
> you meant _patches to nixpkgs_, not `.patch` files _in nixpkgs_.

Well, both actually.

The heavier the patching (both ways), the more maintainers we'd need for the Nix community to accept them.

**Comment by amjoseph-nixpkgs** [comment]
No `patch = ...` entries have been needed so far and I would like to keep it that way.

**Comment by Atemu** [comment]
Yeah, absolutely. I find it highly likely though that you'll need a patch at one point or another.

File: pkgs/applications/editors/emacs/generic.nix [hunk 2]
```diff
@@ -45,6 +44,8 @@
   else if withMotif then "motif"
   else if withAthena then "athena"
   else "lucid")
+, systemdSupport ? stdenv.isLinux
+, systemd
```
Comment (adisbladis) [inline pkgs/applications/editors/emacs/generic.nix:48]:
```suggestion
, systemdSupport ? stdenv.isLinux, systemd
```

This is internally inconsistent styling with other single-package options.

File: pkgs/applications/editors/emacs/generic.nix [hunk 3]
```diff
@@ -45,6 +44,8 @@
   else if withMotif then "motif"
   else if withAthena then "athena"
   else "lucid")
+, systemdSupport ? stdenv.isLinux
```
Comment (adisbladis) [inline pkgs/applications/editors/emacs/generic.nix:47]: Other Emacs options are in the style of `withFoo`, `systemdSupport` is inconsistent with other emacs options.
A better naming would be `withSystemd`.

There is no global consensus in nixpkgs, but at least we can try to be locally consistent.

**Review by adisbladis — CHANGES_REQUESTED** [review]

**Review by Atemu — CHANGES_REQUESTED** [review]
Please squash and update the PR title.

**Comment by amjoseph-nixpkgs** [comment]
Thanks, @adisbladis!

**Pull request merged**